cosmwasm/packages/wasmi-runtime/src/exports.rs

```diff
@@ -54,30 +54,33 @@ pub unsafe extern "C" fn ecall_init(
 ) -> InitResult {
     if let Err(_e) = validate_const_ptr(env, env_len as usize) {
         error!("Tried to access data outside enclave memory!");
-        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall));
+        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall), 0);
     }
     if let Err(_e) = validate_const_ptr(msg, msg_len as usize) {
         error!("Tried to access data outside enclave memory!");
-        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall));
+        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall), 0);
     }
     if let Err(_e) = validate_const_ptr(contract, contract_len as usize) {
         error!("Tried to access data outside enclave memory!");
-        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall));
+        return result_init_success_to_initresult(Err(EnclaveError::FailedFunctionCall), 0);
     }
 
     let contract = std::slice::from_raw_parts(contract, contract_len);
     let env = std::slice::from_raw_parts(env, env_len);
     let msg = std::slice::from_raw_parts(msg, msg_len);
     let result = panic::catch_unwind(|| {
-        let result = crate::wasm::init(context, gas_limit, contract, env, msg);
-        result_init_success_to_initresult(result)
+        let mut used_gas = 0;
+        let result = crate::wasm::init(context, gas_limit, &mut used_gas, contract, env, msg);
+        result_init_success_to_initresult(result, used_gas)
     });
     if let Ok(res) = result {
         res
     } else {
         error!("Call ecall_init panicked unexpectedly!");
         InitResult::Failure {
             err: EnclaveError::Panic,
+            // here we have no choice but to return 0, since the enclave panicked :(
+            used_gas: 0,
```

Comment:
Ok, this is a lie. we could also return the gas limit but then the user will likely observe an "out of gas" which isn't necessarily true, or useful.

What do you think?

Reply:
I vote to return the limit.

Reply:
The limit is set by the caller, no? So returning the limit could charge a whole lot of gas of someone manually set the limit, not knowing an error could cause the entire thing to be charged. If someone was sophisticated they would set a super low gas limit anyway, so this probably wouldn't be a big deterrent. I'd rather keep it at 0 for now, and think about this more as we go

Reply:
seems like we have consensus on the 0. I'll keep it like this for now.